sycl/include/sycl/ext/intel/experimental/device_architecture.hpp

```diff
@@ -0,0 +1,301 @@
+        __SYCL_TARGET_INTEL_GPU_CML__ == 1;
+    arr[static_cast<int>(
+        ext::intel::experimental::architecture::intel_gpu_icllp)] =
+        __SYCL_TARGET_INTEL_GPU_ICLLP__ == 1;
+    arr[static_cast<int>(
+        ext::intel::experimental::architecture::intel_gpu_ehl)] =
+        __SYCL_TARGET_INTEL_GPU_EHL__ == 1;
+    arr[static_cast<int>(
+        ext::intel::experimental::architecture::intel_gpu_tgllp)] =
+        __SYCL_TARGET_INTEL_GPU_TGLLP__ == 1;
+    arr[static_cast<int>(
+        ext::intel::experimental::architecture::intel_gpu_rkl)] =
+        __SYCL_TARGET_INTEL_GPU_RKL__ == 1;
+    arr[static_cast<int>(
+        ext::intel::experimental::architecture::intel_gpu_adl_s)] =
+        __SYCL_TARGET_INTEL_GPU_ADL_S__ == 1;
+    arr[static_cast<int>(
+        ext::intel::experimental::architecture::intel_gpu_rpl_s)] =
+        __SYCL_TARGET_INTEL_GPU_RPL_S__ == 1;
+    arr[static_cast<int>(
+        ext::intel::experimental::architecture::intel_gpu_adl_p)] =
+        __SYCL_TARGET_INTEL_GPU_ADL_P__ == 1;
+    arr[static_cast<int>(
+        ext::intel::experimental::architecture::intel_gpu_adl_n)] =
+        __SYCL_TARGET_INTEL_GPU_ADL_N__ == 1;
+    arr[static_cast<int>(
+        ext::intel::experimental::architecture::intel_gpu_dg1)] =
+        __SYCL_TARGET_INTEL_GPU_DG1__ == 1;
+    arr[static_cast<int>(
+        ext::intel::experimental::architecture::intel_gpu_acm_g10)] =
+        __SYCL_TARGET_INTEL_GPU_ACM_G10__ == 1;
+    arr[static_cast<int>(
+        ext::intel::experimental::architecture::intel_gpu_acm_g11)] =
+        __SYCL_TARGET_INTEL_GPU_ACM_G11__ == 1;
+    arr[static_cast<int>(
+        ext::intel::experimental::architecture::intel_gpu_acm_g12)] =
+        __SYCL_TARGET_INTEL_GPU_ACM_G12__ == 1;
+    arr[static_cast<int>(
+        ext::intel::experimental::architecture::intel_gpu_pvc)] =
+        __SYCL_TARGET_INTEL_GPU_PVC__ == 1;
+  }
+};
+
+// One entry for each enumerator in "architecture" telling whether the AOT
+// target matches that architecture.
+static constexpr IsAOTForArchitectureClass is_aot_for_architecture;
+
+// Reads the value of "is_allowable_aot_mode" via a template to defer triggering
+// static_assert() until template instantiation time.
+template <ext::intel::experimental::architecture... Archs>
+constexpr static bool allowable_aot_mode() {
+  return is_allowable_aot_mode;
+}
+
+// Tells if the current device has one of the architectures in the parameter
+// pack.
+template <ext::intel::experimental::architecture... Archs>
+constexpr static bool device_architecture_is() {
+  return (is_aot_for_architecture.arr[static_cast<int>(Archs)] || ...);
+}
+
+// Helper object used to implement "else_if_architecture_is" and "otherwise".
+// The "MakeCall" template parameter tells whether a previous clause in the
+// "if-elseif-elseif ..." chain was true.  When "MakeCall" is false, some
+// previous clause was true, so none of the subsequent
```

Comment:
Likely nit: I haven't investigated it, but such recursions usually taste C++14'ish to me. Any chance we can restructure using C++17? We might need to return two values but `std::pair` seems to have all the necessary `constexpr`s.

Reply:
Can be applied in separate PR.